File: ios/Packages/GemstoneServices/Sources/Stores/SwapStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import typealias Gemstone.AssetId
import enum Gemstone.GemAssetFilter
import struct Gemstone.GemSwapPair
import protocol Gemstone.GemSwapStore
import enum Gemstone.RecentActivityType
import GemstonePrimitives
import Primitives
import Store

public final class GemstoneSwapStore: GemSwapStore, @unchecked Sendable {
    private let assetStore: AssetStore
    private let transactionStore: TransactionStore
    private let recentActivityStore: RecentActivityStore

    public init(
        assetStore: AssetStore,
        transactionStore: TransactionStore,
        recentActivityStore: RecentActivityStore,
    ) {
        self.assetStore = assetStore
        self.transactionStore = transactionStore
        self.recentActivityStore = recentActivityStore
    }

    public func getSwapPairs(walletId: String) async throws -> [Gemstone.GemSwapPair] {
        try transactionStore.getSwapHistory(walletId: WalletId.from(id: walletId))
            .map { GemSwapPair(fromAssetId: $0.fromAsset.identifier, toAssetId: $0.toAsset.identifier) }
    }

    public func getRecentAssetIds(walletId: String, types: [Gemstone.RecentActivityType], filters: [GemAssetFilter], limit: UInt32) async throws -> [Gemstone.AssetId] {
        try recentActivityStore.getRecent(
            walletId: WalletId.from(id: walletId),
            types: types.map { $0.toPrimitives() },
            limit: Int(limit),
            filters: filters.map { $0.map() },
        ).map(\.asset.id.identifier)
    }

    public func getAssetIds(walletId: String, filters: [GemAssetFilter], limit: UInt32) async throws -> [Gemstone.AssetId] {
        try assetStore.getAssetsData(
            walletId: WalletId.from(id: walletId),
            filters: filters.map { $0.map() },
            limit: Int(limit),
        ).map(\.asset.id.identifier)
    }
}

File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/stores/SwapStore.kt
package com.gemwallet.android.data.services.gemstone.stores

import com.gemwallet.android.data.service.store.database.AssetsDao
import com.gemwallet.android.data.service.store.database.TransactionsDao
import com.gemwallet.android.data.services.gemstone.assets.filteredSearch
import com.gemwallet.android.ext.toPrimitives
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.withContext
import uniffi.gemstone.GemAssetFilter
import uniffi.gemstone.GemSwapPair
import uniffi.gemstone.GemSwapStore
import uniffi.gemstone.RecentActivityType

class GemstoneSwapStore(private val assetsDao: AssetsDao, private val transactionsDao: TransactionsDao) : GemSwapStore {

    override suspend fun getSwapPairs(walletId: String): List<GemSwapPair> = withContext(Dispatchers.IO) {
        transactionsDao.getSwapPairs(walletId).map { GemSwapPair(it.fromAssetId, it.toAssetId) }
    }

    override suspend fun getRecentAssetIds(walletId: String, types: List<RecentActivityType>, filters: List<GemAssetFilter>, limit: UInt): List<String> = withContext(Dispatchers.IO) {
        assetsDao.getRecentAssets(
            walletId = walletId,
            type = types.map { it.toPrimitives() },
            filters = filters.map { it.toRequestFilter() }.toSet(),
            limit = limit.toInt(),
        ).firstOrNull().orEmpty().map { it.asset.id }
    }

    override suspend fun getAssetIds(walletId: String, filters: List<GemAssetFilter>, limit: UInt): List<String> = withContext(Dispatchers.IO) {
        assetsDao.filteredSearch(walletId = walletId, query = "", limit = limit.toInt(), filters = filters.toSet(), withPriority = false).firstOrNull().orEmpty().map { it.id }
    }
}
